File: quick_sort/quick-sort.go
package main

func main() {

}

func quickSort(nums []int, low, high int) []int {
	if low < high {
		nums, pivot := partition(nums, low, high)
		quickSort(nums, low, pivot-1)
		quickSort(nums, pivot+1, high)
	}
	return nums
}

func partition(nums []int, low, high int) ([]int, int) {
	pivot := nums[high]
	i := low

	for j := low; j < high; j++ {
		if nums[j] < pivot {
			nums[i], nums[j] = nums[j], nums[i]
			i++
		}
	}

	nums[high], nums[i] = nums[i], nums[high]
	return nums, i
}

File: quick_sort/quick-sort.ts
export {};

function quickSort(
  nums: Array<number>,
  low: number = 0,
  high: number = nums.length - 1
): number[] {
  if (low < high) {
    const pivot = nums[high];
    let i = low;

    for (let j = low; j < high; j++) {
      if (nums[j] < pivot) {
        [nums[i], nums[j]] = [nums[j], nums[i]];
        i++;
      }
    }

    [nums[i], nums[high]] = [nums[high], nums[i]];
    quickSort(nums, low, i - 1);
    quickSort(nums, low + 1, high);
  }
  return nums;
}
